### PR TITLE
SY-4443: (4/6) Hardware Framework Task and Device Tests

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-4443](https://linear.app/synnax/issue/SY-4443)

## Description

Net-new unit coverage for the hardware framework domains `platform/task` and `platform/device`. Pure additions. **43 files, +3,200.**

## Stack

This is part of the split of #2554 (`SY-4443: Convert Console to Layered Architecture`) into a reviewable, independently-building stack. Merge in order:

1. `sy-4443-1-foundation` → `rc`
2. `sy-4443-2-console-refactor` → (1)
3. `sy-4443-3-tests-platform-core` → (2)
4. `sy-4443-4-tests-hw-framework` → (3)
5. `sy-4443-5-tests-hw-vendors` → (4)
6. `sy-4443-6-tests-visualizations` → (5)

Every branch passes `check-types` and `lint` independently; the tip's tree is identical to the original #2554 branch.

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR is a pure test addition (~3,200 lines across 43 new spec files) providing unit and integration coverage for the `platform/task` and `platform/device` domains of the Console frontend. No production code is changed.

- **`platform/task`**: New specs cover form lifecycle (`wrapForm`, rack-key derivation, unsaved-changes tracking), control-panel interactions (start/stop command dispatch, configure-button, status expand/contract), channel list context-menu actions (enable/disable/remove/duplicate/tare, snapshot-mode suppression), field components (AutoStart, SampleRate, StateUpdateRate, StreamRate, DataSaving), layout patterns (List, DetailsHeader, ListAndDetails), and hooks (`useTare`, `useStatus`, `useKey`, `useIsRunning`, `useIsSnapshot`).
- **`platform/device`**: New specs cover `Configure` modal flow, `ChangeIdentifierMenuItem`, `Select`, `Provider`, `useChangeIdentifier`, schema validators (`nameZ`, `identifierZ`, `commandStatePairZ`), and `checkConfigured`.
- Integration tests consistently use a live `createTestClient()` and assert real side effects (device properties, task creation, tare/start/stop command dispatch via a streamer).

<h3>Confidence Score: 4/5</h3>

Safe to merge — all 43 files are new test specs with no changes to production code.

The tests are well-structured and cover an impressive range of scenarios: form lifecycle, command dispatch via a real streamer, permission gating, snapshot-mode suppression, and schema validation. Two minor test-quality issues exist: `useKey.spec.ts` uses a `?? null` coercion that masks the actual return type, and `createIngester.spec.ts` asserts `placeLayout` synchronously after a `void`-discarded call, which would silently misfire if the implementation ever defers that call past an `await`.

`console/src/platform/task/createIngester.spec.ts` and `console/src/platform/task/useKey.spec.ts` have minor test-precision issues worth a second look.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| console/src/platform/task/Form.spec.tsx | Comprehensive tests for `wrapForm`, `useIsRunning`, and `useIsSnapshot`; covers name-field editing, unsaved-changes tracking, rack-key derivation from layout args, and a live-cluster task creation round-trip. |
| console/src/platform/task/controls/Controls.spec.tsx | Integration tests for `Controls.Controls`; uses a module-level rack created via top-level `await`, verifies start/stop command dispatch via a real streamer, and correctly guards against commands when task key is absent or form is invalid. |
| console/src/platform/task/useTare.spec.tsx | Tests for the `useTare` hook covering `allowTare`, the `tare` command function, and `handleTare` bulk dispatch; integration cases use a real streamer to verify the tare command args. |
| console/src/platform/task/createIngester.spec.ts | Tests for `createIngester`; the happy-path test uses `void ingest(...)` and asserts `placeLayout` synchronously, which is fragile if the implementation ever defers the call past an `await`. |
| console/src/platform/task/useKey.spec.ts | Minimal tests for `useKey`; the "absent" case uses `?? null` coercion before `toBeNull()`, which passes for both `undefined` and `null` and doesn't pin the actual return type. |
| console/src/platform/device/Configure.spec.tsx | Integration tests for the `Configure` modal covering name validation, identifier validation, and a full configure-then-close round-trip against a live cluster. |
| console/src/platform/device/ChangeIdentifierMenuItem.spec.tsx | Tests for `ChangeIdentifierMenuItem`: permission gating, multi-select suppression, unconfigured-device suppression, identifier rename round-trip, and dismiss-without-change. |
| console/src/platform/task/ChannelList.spec.tsx | Tests for `ChannelList` including context-menu actions (enable, disable, remove, duplicate, tare) and snapshot-mode hiding of all mutating actions. |
| console/src/platform/task/export.spec.ts | Tests for `Task.extract` covering direct numeric key, resolution from layout args, error when key maps to neither a task nor a layout, and error when layout has no configured task key. |
| console/src/platform/task/layouts/ListAndDetails.spec.tsx | Tests for the `ListAndDetails` layout: empty state with disabled details, selection-driven panel switching, copy-source forwarding on Add, context-menu duplicate, and context-menu remove with panel clearing. |
| console/src/platform/device/types.spec.ts | Pure unit tests for `nameZ`, `identifierZ`, `commandStatePairZ`, and `ZERO_COMMAND_STATE_PAIR` schemas; all boundary cases (min/max length, invalid start character) are covered. |
| console/src/platform/task/UtilityButtons.spec.tsx | Tests for `UtilityButtons` covering button visibility gating on task key presence and correct clipboard content for JSON config, TypeScript snippet, Python snippet, and shareable link. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[renderWithConsole] --> B[renderInTaskForm]
    A --> C[renderWithDeviceForm]
    B --> D[renderInTaskFormWithClient]
    B --> E[renderTaskFormHook]
    C --> F[renderMenuItem]
    D --> G[Integration tests\nwith live cluster]
    E --> H[Hook unit tests\nno network]
    F --> I[Menu item tests\nwith/without client]
    G -->|creates| J[racks / tasks / channels / devices]
    G -->|opens| K[StreamerClient\nfor command verification]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[renderWithConsole] --> B[renderInTaskForm]
    A --> C[renderWithDeviceForm]
    B --> D[renderInTaskFormWithClient]
    B --> E[renderTaskFormHook]
    C --> F[renderMenuItem]
    D --> G[Integration tests\nwith live cluster]
    E --> H[Hook unit tests\nno network]
    F --> I[Menu item tests\nwith/without client]
    G -->|creates| J[racks / tasks / channels / devices]
    G -->|opens| K[StreamerClient\nfor command verification]
```

</a>

<sub>Reviews (1): Last reviewed commit: ["SY-4443: Add hardware framework task and..."](https://github.com/synnaxlabs/synnax/commit/e92479a0b72fb3b0136d61ed326f760740362ad7) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41758471)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: console/src/platform/task/useKey.spec.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      Task.useKey(ctx),
+    );
+    expect(result.current.value).toBe("task-123");
+  });
+
+  it("should return null when the key field is absent", async () => {
+    const { result } = await renderTaskFormHook({}, (ctx) => Task.useKey(ctx));
+    expect(result.current.value ?? null).toBeNull();
+  });
+});
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply console/src/platform/task/useKey.spec.ts]: Fixed — `useKey` returns `task.Key | undefined`, so the test now asserts `toBeUndefined()` (and is renamed accordingly) instead of coercing with `?? null`.

File: console/src/platform/task/createIngester.spec.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const placeLayout = vi.fn();
+    const data = { device: "dev-1", sampleRate: 100 };
+    void ingest(data, {
+      layout: { key: "layout-1" },
+      placeLayout,
+      store,
+      client,
+      projectKey: "",
+    });
+    expect(placeLayout).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply console/src/platform/task/createIngester.spec.ts:63]: Leaving as-is: `ingest` places the layout synchronously before any await and the test asserts that contract. Will revisit if it becomes async.

File: console/src/platform/task/useKey.spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (emilbon99) [reply console/src/platform/task/useKey.spec.ts]: Correction to my earlier reply: digging in, `useKey` actually returns `null` (not `undefined`) — the field system's `optional` overload returns `O | null`, and the old `?? null` masked it. More importantly the test was hiding a real gap: a new/unsaved task defaults `key` to the zero value `"0"` (present, not absent), so consumers' `key == null` guards didn't catch it. `useKey` now collapses both absent and `"0"` to `null` with an honest `task.Key | null` return, and the spec asserts all three cases (real key, absent, zero).

**Review by pjdotson — APPROVED** [review]
